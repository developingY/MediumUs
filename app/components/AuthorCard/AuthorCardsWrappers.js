import styled, { css } from 'styled-components';
import { Link } from 'react-router-dom';
import getColor from '../../staticData/colorSets';

const StyledWrapper = styled.div`
  display: flex;
  align-items: center;
  min-width: 0;
`;

const AuthorInfoWrapper = styled.div`
  display: initial;
`;

export const StyledButtonWrapper = styled.span`
  margin-left: 8px;
`;

const infoWrapperBaseStyle = css`
  flex: auto;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
`;

const redingInfoWrapperBaseStyle = css`
  color: #0000008a;
  fill: #0000008a;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const aBaseStyle = css`
  color: ${getColor('black').color};
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  :active,
  :hover {
    color: ${getColor('black').hoverColor};
    text-decoration: underline;
    outline: 0;
  }
`;

const spanBaseStyle = css`
  vertical-align: 1px;
`;

function conditionallyStyledComponents(variation) {
  switch (variation) {
    case 'PublicationHome':
      return {
        StyledInfoWrapper: styled.div`
          ${infoWrapperBaseStyle}
          font-size: 15px;
          margin: 0 0 0 10px;
          line-height: 1.4;
          vertical-align: center;
        `,
        ReadingInfoWrapper: styled.div`
          ${redingInfoWrapperBaseStyle}
        `,
        StyledA: styled(Link)`
          ${aBaseStyle}
          font-size: 16px;
          @media only screen and (max-width: 768px) {
            font-size: 15px;
          }
          color: ${props => getColor(props.theme).color};
        `,
        StyledSpan: styled.span`
          ${spanBaseStyle}
          padding: 0 4.5px;
        `,
      };
    case 'TopicHome':
      return {
        StyledInfoWrapper: styled.div`
          ${infoWrapperBaseStyle}
          font-size: 15.8px;
          margin-left: 16px;
          line-height: 20px;
          margin-top: -1px;
        `,
        ReadingInfoWrapper: styled.div`
          ${redingInfoWrapperBaseStyle}
          line-height: 20px;
        `,
        StyledA: styled(Link)`
          ${aBaseStyle}
        `,
        StyledSpan: styled.span`
          ${spanBaseStyle}
          padding: 0 4px;
        `,
      };
    case 'Home':
      return {
        StyledInfoWrapper: styled.div`
          ${infoWrapperBaseStyle}
          font-size: 15px;
          line-height: 1.4;
        `,
        ReadingInfoWrapper: styled.div`
          ${redingInfoWrapperBaseStyle}
        `,
        StyledA: styled(Link)`
          ${aBaseStyle}
        `,
        StyledSpan: styled.span`
          ${spanBaseStyle}
          padding: 0 4.8px;
        `,
      };
    case 'ArticleTitle':
      return {
        StyledInfoWrapper: styled.div`
          ${infoWrapperBaseStyle}
          font-size: 16px;
          margin-left: 12px;
          line-height: 20px;
        `,
        ReadingInfoWrapper: styled.div`
          ${redingInfoWrapperBaseStyle}
          margin-top: 4px;
        `,
        StyledA: styled(Link)`
          ${aBaseStyle}
        `,
        StyledSpan: styled.span`
          ${spanBaseStyle}
          padding: 0 4px;
        `,
      };
    default:
      throw new Error(`Unknown variation: ${variation}`);
  }
}

export function styledComponents(variation) {
  return {
    StyledWrapper,
    AuthorInfoWrapper,
    ...conditionallyStyledComponents(variation),
  };
}
