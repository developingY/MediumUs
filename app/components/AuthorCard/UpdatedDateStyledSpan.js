import styled from 'styled-components';

const UpdatedDateStyledSpan = styled.span`
  font-size: 15px;
  padding: 10px 10px;
  display: inline-block;
  color: #fff;
`;

export default UpdatedDateStyledSpan;
