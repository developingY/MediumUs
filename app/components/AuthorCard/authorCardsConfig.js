export default function getConfig(variation) {
  switch (variation) {
    case 'CollectionHome':
      return { avatarSize: '36px', avatarDisplay: true, followButton: false };
    case 'TopicHome':
      return { avatarSize: '40px', avatarDisplay: true, followButton: false };
    case 'Home':
      return { avatarSize: '', avatarDisplay: false, followButton: false };
    case 'ArticleTitle':
      return { avatarSize: '48px', avatarDisplay: true, followButton: true };
    default:
      throw new Error(`Unknown variantion: ${variation}`);
  }
}
