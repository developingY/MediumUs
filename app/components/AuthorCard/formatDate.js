export default date => {
  const [, month, day, year] = new Date(date).toDateString().split(' ');
  return { month, day, year };
};
