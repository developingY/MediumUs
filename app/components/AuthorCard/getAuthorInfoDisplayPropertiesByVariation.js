export default function getAuthorInfoDisplayPropertiesByVariation(variation) {
  switch (variation) {
    case 'PublicationHome':
      return {
        avatarSize: '36px',
        isDisplayAvatar: true,
        hasFollowButton: false,
      };
    case 'TopicHome':
      return {
        avatarSize: '40px',
        isDisplayAvatar: true,
        hasFollowButton: false,
      };
    case 'Home':
      return { avatarSize: '', isDisplayAvatar: false, hasFollowButton: false };
    case 'ArticleTitle':
      return {
        avatarSize: '48px',
        isDisplayAvatar: true,
        hasFollowButton: true,
      };
    default:
      throw new Error(`Unknown variation: ${variation}`);
  }
}
