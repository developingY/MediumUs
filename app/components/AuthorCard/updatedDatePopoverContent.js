import styled from 'styled-components';

const UpdatedDatePopoverContent = styled.span`
  font-size: 15px;
  padding: 10px 10px;
  display: inline-block;
  color: #fff;
`;

export default UpdatedDatePopoverContent;
