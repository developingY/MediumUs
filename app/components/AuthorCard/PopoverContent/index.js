/**
 *
 * PopoverContent
 *
 */

import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { OutlinedButton } from '../../Button';
import formatDate from '../formatDate';
import Avatar from '../../Avatar';
import {
  UnorderedList,
  Divider,
  TitleA,
  Subtitle,
  MemberSince,
  Img,
  FollowerSpan,
  NumSpan,
} from './styledListItems';
import { UpWrapper, DownWrapper, TitleWrapper } from './styledWrappers';
import roundToThousand from '../../../utils/roundToThousand';
import { selectUserInfo, selectPublicationInfo } from '../../../selectors';

function PopoverContent(props) {
  const { year, month } = formatDate(props.joinedDate);
  return (
    <UnorderedList>
      <UpWrapper>
        <TitleWrapper>
          <TitleA href={props.headerLink}>{props.header}</TitleA>
          {props.member && (
            <MemberSince>{`Medium member since ${month} ${year}`}</MemberSince>
          )}
          <Subtitle>{props.subHeader}</Subtitle>
        </TitleWrapper>

        {props.imgType === 'avatar' ? (
          <Avatar size="60px" id={props.id} />
        ) : (
          <Img src={props.imgLink} alt={props.imgAlt} />
        )}
      </UpWrapper>
      <Divider />
      <DownWrapper>
        <FollowerSpan>
          Followed by{' '}
          <NumSpan>{roundToThousand(props.followersNumber)}</NumSpan> people
        </FollowerSpan>
        <OutlinedButton
          text="Follow"
          type="outlined"
          size="middle"
          theme={props.theme}
        />
      </DownWrapper>
    </UnorderedList>
  );
}

PopoverContent.propTypes = {
  headerLink: PropTypes.string.isRequired,
  header: PropTypes.string.isRequired,
  subHeader: PropTypes.string,
  joinedDate: PropTypes.string,
  member: PropTypes.bool,
  imgLink: PropTypes.string,
  imgAlt: PropTypes.string,
  followersNumber: PropTypes.number.isRequired,
  imgType: PropTypes.oneOf(['avatar', 'publication']),
  id: PropTypes.string,
  theme: PropTypes.string,

  // writerId: PropTypes.string,
};

function mapStateToProps(state, ownProps) {
  const { id, imgType } = ownProps;
  let componentProps = {};
  if (imgType === 'avatar') {
    // const { author: writerId } = selectArticleAbstract(state, id);
    const {
      link,
      name,
      description,
      member = undefined,
      memberJoinedDate = undefined,
      followers,
    } = selectUserInfo(state, id);
    componentProps = {
      headerLink: link,
      header: name,
      subHeader: description,
      member,
      joinedDate: memberJoinedDate,
      followersNumber: followers,
    };
  } else if (imgType === 'publication') {
    const { link, name, description, logo, followers } = selectPublicationInfo(
      state,
      id,
    );
    componentProps = {
      headerLink: link,
      header: name,
      subHeader: description,
      imgLink: logo,
      imgAlt: name,
      followersNumber: followers,
    };
  }
  return componentProps;
}

export default connect(mapStateToProps)(PopoverContent);
