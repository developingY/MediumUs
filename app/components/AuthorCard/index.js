/**
 *
 * AuthorCards
 *
 */

import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Avatar from '../Avatar';
import { styledComponents, StyledButtonWrapper } from './AuthorCardsWrappers';
import { OutlinedButton } from '../Button';
import StarIcon from '../../staticData/svgIcons/starIcon';
import getAuthorInfoDisplayPropertiesByVariation from './getAuthorInfoDisplayPropertiesByVariation';
import OverlayTrigger from '../OverlayTrigger';
import PopoverContent from './PopoverContent/index';
import requiredIfAllPresent from '../../utils/requiredIfAllPresent';
import formatDate from './formatDate';
import UpdatedDateStyledSpan from './UpdatedDateStyledSpan';
import calcReadingTime from '../../utils/calcReadingTime';
import {
  selectArticleAuthorInfo,
  selectArticleAllInfo,
  selectArticlePublicationInfo,
} from '../../selectors';

function AuthorCard(props) {
  const {
    StyledWrapper,
    StyledInfoWrapper,
    StyledA,
    AuthorInfoWrapper,
    ReadingInfoWrapper,
    StyledSpan,
  } = styledComponents(props.variation);

  const {
    avatarSize,
    isDisplayAvatar,
    hasFollowButton,
  } = getAuthorInfoDisplayPropertiesByVariation(props.variation);

  const {
    hoverEffect,
    authorId,
    authorLink,
    authorName,
    publicationId,
    publicationLink,
    publicationName,
    creationDate,
    wordCount,
    isPremium,
    lastModified,
  } = props;

  const avatarElement = (
    <Link to={authorLink}>
      <div>
        <Avatar
          alt={`Go to the profile of ${authorName}`}
          size={avatarSize}
          id={authorId}
        />
      </div>
    </Link>
  );

  const authorElement = (
    <StyledA to={authorLink} key="name" theme={props.theme}>
      {authorName}
    </StyledA>
  );

  function renderAuthor(element) {
    if (hoverEffect) {
      const popoverContent = (
        <PopoverContent imgType="avatar" id={authorId} theme={props.theme} />
      );
      return (
        <OverlayTrigger
          popoverContent={popoverContent}
          trigger="hover"
          placement="top-bottom"
        >
          {element}
        </OverlayTrigger>
      );
    }
    return element;
  }

  const renderPublication = () => {
    const element = (
      <StyledA to={publicationLink} key="publication" theme={props.theme}>
        {publicationName}
      </StyledA>
    );

    if (hoverEffect) {
      const popoverContent = (
        <PopoverContent
          imgType="publication"
          id={publicationId}
          theme={props.theme}
        />
      );
      return (
        <OverlayTrigger
          popoverContent={popoverContent}
          trigger="hover"
          placement="top-bottom"
        >
          {element}
        </OverlayTrigger>
      );
    }
    return element;
  };

  function renderDateTime() {
    const { year, month, day } = formatDate(creationDate);
    let time = `${month} ${day}`;
    // #TODO try memo or sth else to call it less times
    if (year < new Date().getFullYear()) {
      time = time.concat(`, ${year}`);
    }
    const readingTime = `${calcReadingTime(wordCount)} min read`;
    const element = (
      <ReadingInfoWrapper>
        <time title={time}>{time}</time>
        <StyledSpan>&middot;</StyledSpan>
        <span title={readingTime}>{readingTime}</span>
        {isPremium && <StyledSpan>{StarIcon}</StyledSpan>}
      </ReadingInfoWrapper>
    );
    // if the article is updated and have hover effect on, add hover effect for lastModified
    if (hoverEffect && creationDate !== lastModified) {
      const { month: mon, day: d } = formatDate(lastModified);
      return (
        <OverlayTrigger
          popoverContent={
            <UpdatedDateStyledSpan>
              {`Updated ${mon} ${d}`}
            </UpdatedDateStyledSpan>
          }
          popoverColor="rgb(49, 49, 49)"
          trigger="hover"
          placement="top-bottom"
        >
          {element}
        </OverlayTrigger>
      );
    }

    return element;
  }

  return (
    <StyledWrapper>
      {isDisplayAvatar && renderAuthor(avatarElement)}
      <StyledInfoWrapper>
        <AuthorInfoWrapper>
          {renderAuthor(authorElement)}
          {props.publicationId && props.variation !== 'PublicationHome' && (
            <> in {renderPublication()}</>
          )}
          {hasFollowButton && (
            <StyledButtonWrapper>
              <OutlinedButton
                text="Follow"
                size="small"
                type="outlined"
                theme={props.theme}
              />
            </StyledButtonWrapper>
          )}
        </AuthorInfoWrapper>
        <div>{renderDateTime()}</div>
      </StyledInfoWrapper>
    </StyledWrapper>
  );
}

AuthorCard.propTypes = {
  // *
  // author profile related props
  // *
  authorId: PropTypes.string,
  authorLink: PropTypes.string.isRequired,
  authorName: PropTypes.string.isRequired,

  // *
  // article related props
  // *
  creationDate: PropTypes.string.isRequired,
  lastModified: PropTypes.string,
  wordCount: PropTypes.number.isRequired,
  isPremium: PropTypes.bool.isRequired,
  // *
  // publication related props
  // *
  publicationId: PropTypes.string,
  publicationName: PropTypes.string,
  publicationLink: requiredIfAllPresent(['publication'], 'string'),

  variation: PropTypes.oneOf([
    'Home',
    'PublicationHome',
    'TopicHome',
    'ArticleTitle',
  ]).isRequired,
  hoverEffect: PropTypes.bool,
  // eslint-disable-next-line react/no-unused-prop-types
  id: PropTypes.string.isRequired,
  theme: PropTypes.string,
};

function mapStateToProps(state, ownProps) {
  const { id } = ownProps;
  // const { userInfo, articleInfo } = state.testState[id];
  const { name: authorName, link: authorLink } = selectArticleAuthorInfo(
    state,
    id,
  );
  const {
    author: authorId,
    publication: publicationId,
    link: articleLink,
    creationDate,
    lastModified,
    wordCount,
    premium: isPremium,
  } = selectArticleAllInfo(state, id);

  let publicationName;
  let publicationLink;
  if (publicationId) {
    ({
      name: publicationName,
      link: publicationLink,
    } = selectArticlePublicationInfo(state, id));
  }

  return {
    authorId,
    authorName,
    authorLink,
    articleLink,
    creationDate,
    lastModified,
    wordCount,
    isPremium,
    publicationId,
    publicationName,
    publicationLink,
  };
}

export default connect(mapStateToProps)(AuthorCard);
// example:  <AuthorCard id={articleId} variation="Home" hoverEffect theme="green" />

export { AuthorCard };
