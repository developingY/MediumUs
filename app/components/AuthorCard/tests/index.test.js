/**
 *
 * Tests for AuthorCards
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import 'jest-dom/extend-expect'; // add some helpful assertions
import AuthorCard from '../index';
import getConfig from '../authorCardsConfig';
import authorInfo from '../stories/exampleData';

test('AuthorInfo has desired props', () => {
  expect(authorInfo).toHaveProperty('name');
  expect(authorInfo).toHaveProperty('authorLink');
  expect(authorInfo).toHaveProperty('date');
});

test('authorCardsConfig has desired configuration', () => {
  const authorCardsConfig = getConfig('TopicHome');
  expect(authorCardsConfig).toHaveProperty('avatarSize', '40px');
  expect(authorCardsConfig).toHaveProperty('avatarDisplay', true);
  expect(authorCardsConfig).toHaveProperty('followButton', false);
});

describe('<AuthorCard />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<AuthorCard {...authorInfo} variation="CollectionHome" />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Expect to have authorInfo.name and authorInfo.collection in html', () => {
    const { getByText } = render(
      <AuthorCard {...authorInfo} variation="ArticleTitle" />,
    );
    expect(getByText(authorInfo.name)).toBeInTheDocument();
    expect(getByText(authorInfo.collection)).toBeInTheDocument();
    expect(getByText('Follow')).toBeInTheDocument();
  });

  it('Should render and match the snapshot', () => {
    const { container } = render(
      <AuthorCard {...authorInfo} variation="Home" />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Should have in collection on the page', () => {
    const { container } = render(
      <AuthorCard {...authorInfo} variation="TopicHome" />,
    );
    expect(container.getElementsByTagName('a')[2].innerHTML).toEqual(
      authorInfo.collection,
    );
  });
});
