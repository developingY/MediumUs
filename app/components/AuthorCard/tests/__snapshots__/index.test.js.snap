// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthorCard /> Should render and match the snapshot 1`] = `
<div>
  <div
    class="AuthorCardsWrappers__StyledContainer-sc-1v4o32t-0 hNKbRb"
  >
    <div
      class="AuthorCardsWrappers__StyledTextContainer-sc-1v4o32t-10 cVADbr"
    >
      <div
        class="AuthorCardsWrappers__UpTextWrapper-sc-1v4o32t-1 hjAOpW"
      >
        <a
          class="AuthorCardsWrappers__StyledA-sc-1v4o32t-12 cHTpeG"
          href="./"
        >
          Lisa Armstrong
        </a>
         in 
        <a
          class="AuthorCardsWrappers__StyledA-sc-1v4o32t-12 cHTpeG"
          href="./"
        >
          OneZero
        </a>
      </div>
      <div
        class="AuthorCardsWrappers__DownTextWrapper-sc-1v4o32t-11 bDIlkS"
      >
        <time>
          Nov 21
        </time>
        <span
          class="AuthorCardsWrappers__StyledSpanSpecial-sc-1v4o32t-13 cvgNJb"
        >
          Â·
        </span>
        <span>
          13 min read
        </span>
        <span
          class="AuthorCardsWrappers__StyledSpanSpecial-sc-1v4o32t-13 cvgNJb"
        >
          <svg
            height="15"
            viewBox="0 0 15 15"
            width="15"
          >
            <path
              d="M7.44 2.32c.03-.1.09-.1.12 0l1.2 3.53a.29.29 0 0 0 .26.2h3.88c.11 0 .13.04.04.1L9.8 8.33a.27.27 0 0 0-.1.29l1.2 3.53c.03.1-.01.13-.1.07l-3.14-2.18a.3.3 0 0 0-.32 0L4.2 12.22c-.1.06-.14.03-.1-.07l1.2-3.53a.27.27 0 0 0-.1-.3L2.06 6.16c-.1-.06-.07-.12.03-.12h3.89a.29.29 0 0 0 .26-.19l1.2-3.52z"
            />
          </svg>
        </span>
      </div>
    </div>
     
  </div>
</div>
`;
